File: bench/Reference/ParMap.c
/* C reference: parallel cube of 1..20 via pthreads */
#include <stdio.h>
#include <pthread.h>

#define N 20
static long results[N];

static void* worker(void* arg) {
    long i = (long)arg;
    results[i] = (i + 1) * (i + 1) * (i + 1);
    return NULL;
}

int main(void) {
    pthread_t threads[N];
    for (long i = 0; i < N; i++)
        pthread_create(&threads[i], NULL, worker, (void*)i);
    for (int i = 0; i < N; i++)
        pthread_join(threads[i], NULL);
    printf("[");
    for (int i = 0; i < N; i++) {
        if (i > 0) printf(", ");
        printf("%ld", results[i]);
    }
    printf("]\n");
    return 0;
}

File: bench/Reference/par_map.hs
-- Haskell reference: parallel cube of 1..20
import Control.Parallel.Strategies
import Data.List (intercalate)

main :: IO ()
main = do
    let results = parMap rseq (\x -> x * x * x) [1..20 :: Int]
    putStrLn $ "[" ++ intercalate ", " (map show results) ++ "]"
